torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,268 @@
+#include <queue>
+
+#include <torch/csrc/autograd/functions/accumulate_grad.h>
+#include <torch/csrc/autograd/input_buffer.h>
+#include <torch/csrc/distributed/autograd/context/dist_autograd_container.h>
+#include <torch/csrc/distributed/autograd/engine/dist_engine.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+using torch::autograd::AccumulateGrad;
+using torch::autograd::edge_list;
+using torch::autograd::Engine;
+using torch::autograd::GraphRoot;
+using torch::autograd::GraphTask;
+using torch::autograd::Node;
+using torch::autograd::validate_outputs;
+using torch::autograd::variable_list;
+
+DistEngine::DistEngine()
+    : initializedContextIds_(), engine_(Engine::get_default_engine()) {}
+
+DistEngine& DistEngine::getInstance() {
+  static DistEngine engine;
+  return engine;
+}
+
+void DistEngine::validateRootsAndRetrieveEdges(
+    const variable_list& roots,
+    edge_list& rootEdges,
+    variable_list& grads) {
+  TORCH_CHECK(!roots.empty(), "No tensors provided for gradient computation.");
+  TORCH_INTERNAL_ASSERT(rootEdges.empty());
+  TORCH_INTERNAL_ASSERT(grads.empty());
+
+  // Verify roots are all scalar and require gradients.
+  for (const auto& root : roots) {
```

Comment:
I feel like you could have avoided writing this (duplicate) code if you just generalized the distributed engine to support explicit gradients on input.

Reply:
> I tried to keep the API simple and minimal in this PR. 

Sure, but better not to do this at the cost of having to copy paste a lot of code :)